[daikin_madoka] Add new Daikin Madoka integration

Petapton opened the pull request:
# What does this implement/fix?
New integration added for daikin madoka thermostat devices ([BRC1H](https://www.daikin.eu/en_us/product-group/control-systems/brc1h.html)).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** None

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3463

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
...

esp32_ble_tracker:

esp32_ble:
  io_capability: display_yes_no

ble_client:
  - mac_address: <mac_address>
    id: madoka_climate

climate:
  - platform: daikin_madoka
    name: "Madoka Climate"
    ble_client_id: madoka_climate
    update_interval: 15s

...

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Petapton committed:
Add Daikin Madoka Component.

Petapton committed:
Fix compatibility with esp-idf framework

Petapton committed:
Add support for fan modes in Madoka integration

Petapton committed:
Use more descriptive variable names in Madoka integration

Petapton committed:
Refactor Madoka integration

Petapton committed:
Update CODEOWNERS

Petapton committed:
Replace numeric define with static const variable in Madoka component

Petapton committed:
Apply clang-format style to Madoka component

Petapton committed:
Merge branch 'esphome:dev' into madoka

Petapton committed:
Apply clang-tidy style to Madoka component

Petapton committed:
Apply clang-format style to Madoka component

Petapton committed:
Apply clang-tidy style (for ESP32 Arduino 2/4) to Madoka component

Petapton commented:
Because no one reviewed the PR. @jesserockz would you mind taking a look?

PS: I don't know how polite is pinging maintainers for attention. So, just in case, my apologies.

Frank802 commented:
Come on guys! Could you please review the PR and complete the merge? Thanks 😊

Petapton committed:
Merge remote-tracking branch 'upstream/dev' into madoka

Petapton committed:
Add madoka component test for esp32-idf

jesserockz commented on esphome/components/madoka/madoka.h:
```diff
@@ -0,0 +1,114 @@
+
+#define TO_ESPBTUUID(x) espbt::ESPBTUUID::from_raw(std::string(x))
+
+#define MADOKA_SERVICE_UUID TO_ESPBTUUID("2141e110-213a-11e6-b67b-9e71128cae77")
```
Please avoid the defines here
```suggestion
static const espbt::ESPBTUUID MADOKA_SERVICE_UUID = espbt::ESPBTUUID::from_raw("2141e110-213a-11e6-b67b-9e71128cae77");
```

jesserockz commented on esphome/components/madoka/madoka.h:
```diff
@@ -0,0 +1,114 @@
+    traits.set_supports_current_temperature(true);
+    return traits;
+  }
+  void set_unit_of_measurement(const char *);
```
Not used?

jesserockz commented on esphome/components/madoka/madoka.cpp:
```diff
@@ -0,0 +1,413 @@
+    return;
+  if (call.get_mode().has_value()) {
+    ClimateMode mode = *call.get_mode();
+    std::vector<chunk> pkt;
```
Not used
```suggestion
```

jesserockz commented on esphome/components/madoka/madoka.h:
```diff
@@ -0,0 +1,114 @@
+
+static const char *const TAG = "madoka";
+
+using chunk = std::vector<uint8_t>;
+using message = std::vector<uint8_t>;
```
These really dont help with readability...

Petapton replied:
What do you suggest? Just use one, or remove them both?

jesserockz commented on esphome/components/madoka/madoka.cpp:
```diff
@@ -0,0 +1,413 @@
+    }
+    ESP_LOGD(TAG, "status: %d, mode: %d", status_out, mode_out);
+    if (mode_out != 255) {
+      this->query_(0x4030, message({0x20, 0x01, (uint8_t) mode_out}), 600);
```
`0x4030` etc should be named static consts at the top of this file

jesserockz commented on esphome/components/madoka/madoka.cpp:
```diff
@@ -0,0 +1,413 @@
+    esp_err_t status;
+    for (int j = 0; j < BLE_SEND_MAX_RETRIES; j++) {
+      status = esp_ble_gattc_write_char(this->parent_->get_gattc_if(), this->parent_->get_conn_id(), this->wwr_handle_,
+                                        chk.size(), &chk[0], ESP_GATT_WRITE_TYPE_NO_RSP, ESP_GATT_AUTH_REQ_NONE);
```
```suggestion
                                        chk.size(), chk.data(), ESP_GATT_WRITE_TYPE_NO_RSP, ESP_GATT_AUTH_REQ_NONE);
```

jesserockz commented on esphome/components/madoka/madoka.cpp:
```diff
@@ -0,0 +1,413 @@
+  }
+}
+
+std::vector<chunk> Madoka::split_payload_(message msg) {
+  std::vector<chunk> result;
+  size_t len = msg.size();
+  result.push_back(chunk({0x00, (uint8_t) (len + 1)}));
+  result[0].insert(result[0].end(), msg.begin(), min(msg.begin() + (MAX_CHUNK_SIZE - 2), msg.end()));
+  int i = 0;
+  for (i = 1; i < len / (MAX_CHUNK_SIZE - 1); i++) {  // from second to second-last
+    result.emplace_back(msg.begin() + ((MAX_CHUNK_SIZE - 1) * i - 1),
+                        msg.begin() + ((MAX_CHUNK_SIZE - 1) * (i + 1) - 1));
+  }
+  if (len > 18) {
+    i++;
+    result.emplace_back(msg.begin() + ((MAX_CHUNK_SIZE - 1) * i), msg.end());
+  }
+  return result;
```
This feels kinda messy and hard to understand where the splits are. Can this be cleaned up and less repetitive? splits are added into `result` in 3 different places where there should be a single loop only that handles the full length. (unless i am missing something [could be due to readability])

jesserockz commented on esphome/components/madoka/madoka.cpp:
```diff
@@ -0,0 +1,413 @@
+  uint8_t message_size = msg.size();
+
+  switch (function_id) {
+    case 0x0020:
```
Named static consts please

```suggestion
    case 0x0020:
```

Uzilahav commented:
HI
Is there a step by step instruction how to install this on ESP32 ?
I"m trying to install it and dont know where to start 
Thanks

Petapton commented:
> HI Is there a step by step instruction how to install this on ESP32 ? I"m trying to install it and dont know where to start Thanks

Yes, there's something on a thread on the home assistant forum. Here's the [link](https://community.home-assistant.io/t/daikin-madoka-brc1h-bluetooth-custom-component/268594/104) to a message with a sample configuration.

Petapton committed:
Keep up with latest changes

Petapton committed:
Move UUID defines to static const variables

Petapton committed:
Remove unused `set_unit_of_measurement`

Petapton committed:
Replace `chunk` and `message` with the more explicit `std::vector<uint8_t>`

Petapton committed:
Use .data() instead of &[0]

Petapton committed:
Use named static const for protocol commands

Petapton committed:
Improve code quality and consistency

Petapton committed:
Refactor `split_payload_`

Petapton committed:
Keep up with latest changes

ToniRos commented:
Hi, after upgrading ESPHome to version 2025.6.0, the Madoka component stops working if it is recompiled.

Petapton committed:
Add auth-related GAP events to the list of queued event types

Petapton commented:
> Hi, after upgrading ESPHome to version 2025.6.0, the Madoka component stops working if it is recompiled.

Until this is sorted out, you could use the following snippet in config:
```
external_components:
  - source: github://Petapton/esphome@madoka
    components: [ madoka, esp32_ble ]
```

Petapton committed:
Revert last commit since a better solution is already available in upstream

Petapton committed:
Import changes to esp32_ble

Petapton commented:
> > Hi, after upgrading ESPHome to version 2025.6.0, the Madoka component stops working if it is recompiled.
> 
> It seems to be related to #9052: by limiting the type of events that get added to the queue, we can't get auth-related events anymore, hence no pairing can happen. I'm working on it.

The issue has been already solved in #9138. While these changes are released, you could use the following in your configuration:
```
external_components:
  - source: github://Petapton/esphome@madoka
    components: [ madoka, esp32_ble ]
```

Petapton committed:
Change `climate.CLIMATE_SCHEMA.extend to climate.climate_schema`

Petapton committed:
Keep up with latest changes

Petapton committed:
fix call to climate_schema

ToniRos commented:
Tested and working perfectly. Thanks.

Petapton committed:
Fix log message for duplicate chunk ID

Petapton committed:
Fix command handling in madoka.cpp

Add missing break statements

swoboda1337 reviewed (changes requested):
I reopened the docs pr. It does need to be updated though as the docs have moved to md from rst. 

There are three other daikin components. Does it make sense to call this daikin_madoka to match the others?

Thanks!

Petapton commented:
Yes, it might make sense.
I guess I named it like this because those are IR profiles and I didn't want to confuse things. I don't actually remember exactly, but I'd have no problem renaming it.

On the other hand, this would require existing users to migrate their configurations. So, if it has to be done, the sooner the better.
Just let me know.

Petapton committed:
Rename madoka component to daikin_madoka

Petapton committed:
Rename esp32 test

Petapton committed:
Add test config for esp32-c3

Petapton committed:
Update CODEOWNERS

swoboda1337 commented on tests/components/daikin_madoka/test.esp32-c3-idf.yaml:
```diff
@@ -0,0 +1 @@
+<<: !include common.yaml
```
Does this need to be updated to the new test format? Look some other tests with ble_client: https://github.com/esphome/esphome/blob/dev/tests/components/pvvx_mithermometer/test.esp32-c3-idf.yaml

amnsharif committed:
Refactor temperature support traits to use feature flags

Petapton committed:
Keep up with latest changes

Petapton committed:
Refactor temperature support traits to use feature flags

Merge pull request #2 from amnsharif/patch-1

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

Petapton committed:
Keep up with latest changes

amnsharif committed:
Refactor logging to use addr variable

Petapton committed:
Keep up with latest changes

Petapton committed:
Fix build issues in 2025.12

BLEClientBase::address_str() now returns char * instead of std::string
Merge pull request #4 from amnsharif/patch-1

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

Petapton committed:
Keep up with latest changes

Petapton committed:
Refactor temperature setpoint handling in DaikinMadoka control method to allow for single setpoint requests

Petapton committed:
Format passkey notification log in gap_event_handler for better readability

Petapton committed:
Fix clang-tidy complains

Petapton committed:
Keep up with latest changes

Petapton committed:
Fix clang-tidy complaints

Petapton commented on esphome/components/daikin_madoka/daikin_madoka.cpp:
```diff
@@ -0,0 +1,440 @@
+    ESP_LOGI(TAG, "Chunk discarded: invalid length.");
+    return;
+  }
+  uint8_t chunk_id = chk[0];
+  std::vector<uint8_t> stripped{chk.begin() + 1, chk.end()};
+  if (chunk_id == 0 && validate_buffer(stripped)) {
+    this->parse_cb_(stripped);
+    return;
+  }
```
Kind of an arbitrary policy here. Adding a warning to monitor the behavior in addition.

Petapton commented on esphome/components/daikin_madoka/daikin_madoka.cpp:
```diff
@@ -0,0 +1,440 @@
+      return;
+    }
+  }
+  delay(t_d);
```
Must defer this for a future refactor

Petapton committed:
Apply copilot review suggestions

Petapton committed:
refactor(daikin_madoka): replace blocking delay() with non-blocking set_timeout() chain

Petapton committed:
refactor(daikin_madoka): centralize command cooldown management

mikecrawford99 commented:
@swoboda1337  @jesserockz sorry to ping. This PR has now been open **for over 2.5 years**, and since my last comment the Copilot review was requested, feedback was addressed in follow-up commits, and the PR still appears to be waiting on maintainer review.

Could you please either take it through review, or leave a short list of remaining blockers so @Petapton and/or the rest of us can help get it finished?

swoboda1337 commented:
@esphbot review
